sql: add metric `sql.failure.count`

vilterp opened the pull request:
Counts the number of queries which result in an error.

Touches #14148

I am not sure whether this interacts with retries in the right way (or even what the right way is). It seems that this would result in the counter getting incremented once for every time a statement is retried and fails. Maybe that's ok, or maybe it should count as a single failure.

Commit:
sql: add metric `sql.failure.count`

Counts the number of queries which result in an error.

Release note (sql change): introduce metric `sql.failure.count`, which
counts the number of queries that result in an error.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26731)
<!-- Reviewable:end -->

andreimatei commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/conn_executor.go, line 1524 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26731#-LEzX09n59-5Jp3A3sSO:-LEzX09n59-5Jp3A3sSP:bv8uc1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdb6dd69126af0a2b4da72247c3c311aa1df0cab/pkg/sql/conn_executor.go#L1524)):*
> ```Go
> // eventNonRetriableErr, depending on the error type.
> func (ex *connExecutor) makeErrEvent(err error, stmt tree.Statement) (fsm.Event, fsm.EventPayload) {
> 	ex.server.StatementCounters.FailureCount.Inc(1)
> ```

I think this is not the most appropriate place to put this. This method is called in all sorts of circumstances.
I'd probably put it after the call to `execStmtInOpenState()` - looking at the returned event

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26731)*
<!-- Sent from Reviewable.io -->

couchand commented:
maybe in `recordStatementSummary`?

vilterp commented:
Thanks for the quick review! Moved the increment and added a test.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/conn_executor.go, line 1524 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26731#-LEzX09n59-5Jp3A3sSO:-LEzsJQ-1xBRARBNEhZL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdb6dd69126af0a2b4da72247c3c311aa1df0cab/pkg/sql/conn_executor.go#L1524)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this is not the most appropriate place to put this. This method is called in all sorts of circumstances.
I'd probably put it after the call to `execStmtInOpenState()` - looking at the returned event
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26731#-:-LEzsLo_9gnDilbn2CWP:b-d7e8e3)*
<!-- Sent from Reviewable.io -->

vilterp commented:
@couchand putting it in `recordStatementSummary` would miss planning errors (the kind Dan originally wanted to see in his issue), since we bail out of `dispatchToExecutionEngine` before calling to `recordStatementSummary` if there is a planning error: https://github.com/cockroachdb/cockroach/blob/57b1262b830c7cd417c4cf4c2f703295078e9520/pkg/sql/conn_executor_exec.go#L599-L602

I put it higher up the call stack so it'll get called whether there's a planning or runtime error.

There is an argument to be made that `recordStatementSummary` should similarly be moved up, but then I don't know what we'd do with its metrics like run latency, etc.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26731#-:-LEzuh2REXbhSxLsKUdF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vilterp commented:
Would be nice to break planning and runtime errors into separate metrics, but this seems better than nothing.

bors r+
